node/subsystem/src/jaeger.rs

```diff
@@ -0,0 +1,256 @@
+#[inline(always)]
+pub fn hash_span(hash: &Hash, span_name: impl Into<String>) -> JaegerSpan {
+	INSTANCE.lock().span(hash, span_name).into()
+}
+
+/// Stateful convenience wrapper around [`mick_jaeger`].
+pub enum Jaeger {
+	/// Launched and operational state.
+	Launched {
+		/// [`mick_jaeger`] provided API to record spans to.
+		traces_in: Arc<mick_jaeger::TracesIn>,
+	},
+	/// Preparation state with the necessary config to launch the collector.
+	Prep(JaegerConfig),
+	/// Uninitialized, suggests wrong API usage if encountered.
+	None,
+}
+
+impl Jaeger {
+	/// Spawn the jaeger instance.
+	pub fn new(cfg: JaegerConfig) -> Self {
+		Jaeger::Prep(cfg)
+	}
+
+	/// Spawn the background task in order to send the tracing information out via udp
+	pub fn launch(self) -> result::Result<(), JaegerError> {
+		let cfg = match self {
+			Self::Prep(cfg) => Ok(cfg),
+			Self::Launched{ .. } => {
+				return Err(JaegerError::AlreadyLaunched)
+			}
+			Self::None => Err(JaegerError::MissingConfiguration),
+		}?;
+
+		let jaeger_agent = cfg.agent_addr;
+
+		log::info!("🐹 Collecting jaeger spans for {:?}", &jaeger_agent);
+
+		let (traces_in, mut traces_out) = mick_jaeger::init(mick_jaeger::Config {
+			service_name: format!("{}-{}", cfg.node_name, cfg.node_name),
+		});
+
+
+		// Spawn a background task that pulls span information and sends them on the network.
+		let _handle = async_std::task::spawn::<_, result::Result<(), JaegerError>>(async move {
+			let mut udp_socket = async_std::net::UdpSocket::bind("127.0.0.1:34254").await;
+
+			let mut port = 49000_u16;
+			while udp_socket.is_err() && port < std::primitive::u16::MAX {
+				udp_socket = async_std::net::UdpSocket::bind(format!("127.0.0.1:{}", port)).await;
+				port += 1;
+			}
+			let udp_socket = udp_socket.map_err(|e| JaegerError::PortAllocationError(e))?;
+
+			loop {
+				let buf = traces_out.next().await;
+				// UDP sending errors happen only either if the API is misused (in which case
+				// panicking is desirable) or in case of missing privilege.
+				if let Err(e) = udp_socket.send_to(&buf, jaeger_agent).await
+					.map_err(|e| JaegerError::SendError(e))
+				{
+					log::trace!("Failed to send jaeger span: {:?}", e);
+				}
+			}
+		});
+
+		
+		*INSTANCE.lock() =Self::Launched {
+			traces_in,
+		};
+		Ok(())
+	}
+
+	#[inline(always)]
+	fn traces_in(&self) -> Option<&Arc<mick_jaeger::TracesIn>> {
+		match self {
+			Self::Launched {
+				traces_in,
+				..
+			} => Some(&traces_in),
+			_ => None,
+		}
+	}
+
+	fn span(&self, hash: &Hash, span_name: impl Into<String>) -> Option<mick_jaeger::Span> {
+		if let Some(traces_in) = self.traces_in() {
+			let trace_id = {
+				let mut buf = [0u8; 16];
+				buf.copy_from_slice(&hash.as_ref()[0..16]);
+				std::num::NonZeroU128::new(u128::from_be_bytes(buf))
+			}.expect("16 bytes make a u128; qed");
+			Some(traces_in.span(trace_id, span_name))
+		} else {
+			None
+		}
+	}
+}
+
+lazy_static::lazy_static! {
```

Comment:
Purely as a style preference, I like to see static declarations near the top of the file.